{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8fa3f109",
   "metadata": {
    "id": "8fa3f109"
   },
   "source": [
    "# Custom filtering and clustering\n",
    "\n",
    "This notebook shows how you can do custom filtering and clustering.\n",
    "\n",
    "There are two ways to run this notebook:\n",
    "\n",
    "- As a web server for **biologists**, using the example filters and clustering methods provided here.\n",
    "\n",
    "- As a starting point for **bioinformaticians**, to create their own filters and/or clustering methods.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "Aii5cw4jpq_2",
   "metadata": {
    "id": "Aii5cw4jpq_2"
   },
   "source": [
    "## Instructions for biologists (non-programmers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "249fdaaf",
   "metadata": {
    "id": "249fdaaf"
   },
   "source": [
    "You should run this notebook in [Google Colab](https://colab.google.com).\n",
    "\n",
    "The notebook will download the ProtNAff RNA fragment library (trinucleotides), then apply a filtering step, and finally cluster the filtered fragments. \n",
    "In the *User parameters* section below, choose:\n",
    "\n",
    "- A filtering method.\n",
    "\n",
    "- A clustering method. \n",
    "\n",
    "- A clustering RMSD threshold (in angstroms).\n",
    "\n",
    "- A secondary clustering RMSD threshold. This threshold is used only by the hierarchical method and should be slightly (~1 angstrom) higher than the primary threshold.\n",
    "\n",
    "The filtering methods should be considered as proof-of-principle examples, to be extended by bioinformaticians using this notebook. The filters select purely single-stranded RNA fragments and/or fragments from NMR structures.\n",
    "\n",
    "Three clustering methods are available. \n",
    "\n",
    "The \"full matrix\" method gives the fewest clusters, but is limited to 20 000 fragments. \n",
    "\n",
    "The \"fast\" method gives more clusters, but has no limitation. \n",
    "\n",
    "The \"hierarchical\" method combines the two, and is very close to the approach in the main ProtNAff pipeline.\n",
    "\n",
    "**After filling in the parameters, in the Colab menu, select *Runtime => Run all cells***\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hM9WF2xSp5c3",
   "metadata": {
    "id": "hM9WF2xSp5c3"
   },
   "source": [
    "## Instructions for bioinformaticians"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16b0e681",
   "metadata": {
    "id": "16b0e681"
   },
   "source": [
    "You can run this notebook either in [Google Colab](https://colab.google.com) or under your own Jupyter instance.\n",
    "\n",
    "In this notebook:\n",
    "\n",
    "- A number of filters are defined, and cataloged in the \"filters\" dictionary.\n",
    "\n",
    "- A number of clustering methods are defined, and cataloged in the \"clustering_methods\" dictionary.\n",
    "\n",
    "\n",
    "### You can select a filter and a clustering method and apply it on the ProtNAff fragment library.\n",
    "\n",
    "In \"User parameters\", you can select:\n",
    "- One filter, among the cataloged filters.\n",
    "- One clustering method, among the cataloged clustering methods.\n",
    "- The clustering RMSD threshold.\n",
    "- A secondary clustering RMSD threshold (if hierarchical clustering is chosen)\n",
    "\n",
    "The ProtNAff RNA fragment library is downloaded and loaded in memory. \n",
    "\n",
    "Filtering and clustering is then performed on the fragment library.\n",
    "\n",
    "The number of clusters per motif is shown, and the size distribution of the clusters is plotted.\n",
    "\n",
    "\n",
    "### You can create your own filter\n",
    "\n",
    "- The example filters are based on mask functions. Two example mask functions are provided: one to select NMR fragments, and one to select only purely single-stranded RNA fragments. You can use these examples to write your own mask functions.\n",
    "\n",
    "- The example filter functions show how to invoke and combine mask functions. You can use this to write your own filter function. Catalog your filter function in the \"filters\" dictionary.\n",
    "\n",
    "### You can create your own clustering\n",
    "\n",
    "- Three example clustering functions are provided, based on the ones used in the ProtNAff pipeline. You can write your own clustering function based on this. A clustering function is expected to designate some of the clustered fragments as \n",
    "- The downloaded fragment library coordinates contain trinucleotide xyz values, fitted on a common reference (the first structure), with clashing fragments removed."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "043d6929",
   "metadata": {
    "id": "043d6929"
   },
   "source": [
    "# Definition of example filters and clustering methods"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58440641",
   "metadata": {
    "id": "58440641"
   },
   "source": [
    "## Fragment library definition (before loading)\n",
    "\n",
    "These are global variables used in the filters. \"motifs\" is also used in clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9298d7e0",
   "metadata": {
    "cellView": "form",
    "id": "9298d7e0"
   },
   "outputs": [],
   "source": [
    "#@title\n",
    "motifs = [\"AAA\", \"AAC\", \"ACA\", \"ACC\", \"CAA\", \"CAC\", \"CCA\", \"CCC\"]\n",
    "try:\n",
    "    structures\n",
    "except NameError:\n",
    "    structures = None\n",
    "    fragments = None\n",
    "    coordinates = None\n",
    "    rev_fragindex = None  # to map a key in fragments dict to an index in the fragment array"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1249f5e5",
   "metadata": {
    "id": "1249f5e5"
   },
   "source": [
    "## Masks for filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "783dcd18",
   "metadata": {
    "cellView": "form",
    "id": "783dcd18"
   },
   "outputs": [],
   "source": [
    "#@title\n",
    "def mask_pure_ss(motif_fragments,structures, motif_rev_fragindex):\n",
    "    \"\"\"Returns a mask that indicates for each fragment \n",
    "if it is purely single-stranded or not\"\"\"\n",
    "    ss_set = set([\"L\", \"T\", \"S\", \"J\", \"B\", \"I\"])\n",
    "    \n",
    "    reverse_mapping = {}\n",
    "    for strucname, structure in structures.items():\n",
    "        struc_rev_mapping = {}\n",
    "        for chain, mapping in structure[\"mapping\"].items():\n",
    "            struc_rev_mapping[chain] = {v:k for k,v in mapping.items()}\n",
    "        reverse_mapping[strucname] = struc_rev_mapping\n",
    "    \n",
    "    result = np.zeros(len(motif_rev_fragindex),bool)\n",
    "\n",
    "    for fragkey, frag in motif_fragments.items():\n",
    "        if fragkey not in motif_rev_fragindex:\n",
    "            # fragment contains clashes\n",
    "            continue\n",
    "        fragind = motif_rev_fragindex[fragkey]\n",
    "        structure = frag[\"structure\"]\n",
    "        chain = frag[\"chain\"]\n",
    "        rmap = reverse_mapping[structure][\"chain_\" + chain]\n",
    "        res = [rmap[resid] for resid in frag[\"resid\"]]\n",
    "        ss_states = structures[structure][\"ss\"][\"chain_\" + chain]\n",
    "        frag_ss_states = [ss_states[\"res_\"+ r][0] for r in res]\n",
    "        is_pure_ss = all([ss_state in ss_set for ss_state in frag_ss_states])\n",
    "        result[fragind] = is_pure_ss\n",
    "\n",
    "    return result\n",
    "\n",
    "\n",
    "def mask_nmr(motif_fragments,structures, motif_rev_fragindex):\n",
    "    \"\"\"Returns a mask that indicates for each fragment \n",
    "if it comes from NMR or not\"\"\"\n",
    "    result = np.zeros(len(motif_rev_fragindex),bool)\n",
    "\n",
    "    for fragkey, frag in motif_fragments.items():\n",
    "        if fragkey not in motif_rev_fragindex:\n",
    "            # fragment contains clashes\n",
    "            continue\n",
    "        fragind = motif_rev_fragindex[fragkey]\n",
    "        structure = frag[\"structure\"]\n",
    "        struc = structures[structure]\n",
    "        is_nmr = (struc[\"method\"].find(\"nmr\") > -1)\n",
    "        result[fragind] = is_nmr\n",
    "    \n",
    "    return result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62798003",
   "metadata": {
    "id": "62798003"
   },
   "source": [
    "## Mask-based filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c535aa30",
   "metadata": {
    "cellView": "form",
    "id": "c535aa30"
   },
   "outputs": [],
   "source": [
    "#@title\n",
    "def filter_pure_ss():\n",
    "    result = {}\n",
    "    for motif in fragments:\n",
    "        frags = fragments[motif]\n",
    "        mask = mask_pure_ss(frags, structures, rev_fragindex[motif])\n",
    "        result[motif] = coordinates[motif][mask]\n",
    "    return result\n",
    "\n",
    "def filter_pure_ss_nmr():\n",
    "    result = {}\n",
    "    for motif in fragments:\n",
    "        frags = fragments[motif]\n",
    "        args = frags, structures, rev_fragindex[motif]\n",
    "        mask1 = mask_pure_ss(*args)\n",
    "        mask2 = mask_nmr(*args)\n",
    "        mask = mask1 & mask2\n",
    "        result[motif] = coordinates[motif][mask]\n",
    "    return result\n",
    "\n",
    "def filter_pure_ss_not_nmr():\n",
    "    result = {}\n",
    "    for motif in fragments:\n",
    "        frags = fragments[motif]\n",
    "        args = frags, structures, rev_fragindex[motif]\n",
    "        mask1 = mask_pure_ss(*args)\n",
    "        mask2 = mask_nmr(*args)\n",
    "        mask = mask1 & (~mask2)\n",
    "        result[motif] = coordinates[motif][mask]\n",
    "    return result\n",
    "\n",
    "def filter_none():\n",
    "    return coordinates\n",
    "\n",
    "filters = {\n",
    "    \"None\": filter_none,\n",
    "    \"Pure single-stranded\": filter_pure_ss,\n",
    "    \"Pure single-stranded, only NMR structures\": filter_pure_ss_nmr,\n",
    "    \"Pure single-stranded, no NMR structures\": filter_pure_ss_not_nmr,\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9442a152",
   "metadata": {
    "id": "9442a152"
   },
   "source": [
    "## Clustering methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1e4f4757",
   "metadata": {
    "cellView": "form",
    "id": "1e4f4757"
   },
   "outputs": [],
   "source": [
    "#@title\n",
    "def reassign(structures, clustids, chunksize):\n",
    "    \"\"\"Reassigns structures to the closest cluster center.\n",
    "    Cluster centers are supplied in the form of cluster IDs, i.e.\n",
    "    a list of integer IDs where each ID is the index of the cluster center.\n",
    "    For example, [0, 3, 5] defines three cluster centers, the 1st 4th and 5th structure\n",
    "\n",
    "    chunksize: number of structures to put in a chunk\n",
    "        This is an implementation detail that only affects the speed, not the result\n",
    "    \"\"\"\n",
    "    if len(structures.shape) == 3:\n",
    "        assert structures.shape[2] == 3\n",
    "        structures = structures.reshape(structures.shape[0], structures.shape[1]*3)\n",
    "    if len(structures.shape) == 2:\n",
    "        assert structures.shape[1] % 3 == 0\n",
    "\n",
    "    clusters = {a:[a] for a in clustids}\n",
    "    clus = structures[clustids]\n",
    "    for n in range(0, len(structures), chunksize):\n",
    "        chunk = structures[n:n+chunksize]\n",
    "        d = chunk[:, np.newaxis, :] - clus[np.newaxis, :, :]\n",
    "        inter_sd = np.einsum(\"...ij,...ij->...i\", d,d)\n",
    "        best = np.argmin(inter_sd, axis=1)\n",
    "        for nn in range(len(chunk)):\n",
    "            bestclust = clustids[best[nn]]\n",
    "            if bestclust == (n+nn):\n",
    "                continue\n",
    "            clusters[bestclust].append(n+nn)\n",
    "    return clusters\n",
    "\n",
    "def cluster(structures, threshold, **kwargs):\n",
    "    \"\"\"Clusters structures using an RMSD threshold\n",
    "        First, calculates the full pairwise RMSD matrix.\n",
    "        The maximum number of structures is fixed at 20 000 (400 million pairwise RMSDs)\n",
    "        \n",
    "        The structure with the largest number of other structures within the RMSD\n",
    "         threshold becomes a cluster.\n",
    "        That structure and all of those other structures are removed.\n",
    "        The process is repeated until no structures are left.\n",
    "\n",
    "    This algorithm is used in ATTRACT and in the ProtNAff pipeline \n",
    "    (./create_frag_library/cluster-npy.py)\n",
    "    \"\"\"\n",
    "    if len(structures.shape) == 3:\n",
    "        assert structures.shape[2] == 3\n",
    "        structures = structures.reshape(structures.shape[0], structures.shape[1]*3)\n",
    "    if len(structures.shape) == 2:\n",
    "        assert structures.shape[1] % 3 == 0\n",
    "\n",
    "    if len(structures) > 20000:\n",
    "        raise ValueError(\"Number of structures is too large, maximum is 20 000\")\n",
    "\n",
    "    dist = pdist(structures, 'sqeuclidean')\n",
    "    d = squareform(dist)\n",
    "    lim = threshold*threshold*structures.shape[1]/3\n",
    "    d2 = d<lim\n",
    "    del d\n",
    "\n",
    "    clustids = []\n",
    "    clustered = 0\n",
    "    while clustered < len(structures):\n",
    "        neigh = d2.sum(axis=0)\n",
    "        cluster_center = neigh.argmax()\n",
    "        clustids.append(cluster_center)\n",
    "        cluster = np.where(d2[cluster_center])[0]\n",
    "        for cs in cluster:\n",
    "            d2[cs,:] = False\n",
    "            d2[:, cs] = False\n",
    "        clustered += len(cluster)\n",
    "    return clustids\n",
    "\n",
    "def fastcluster(structures, threshold, chunksize, existing=[], **kwargs):\n",
    "    \"\"\"Clusters structures using an RMSD threshold\n",
    "        First structure becomes a cluster,\n",
    "         second structure only if it doesn't cluster with the first, etc.\n",
    "\n",
    "        structures: 2D numpy array, second dimension = 3 * natoms\n",
    "          structures must already have been fitted!\n",
    "        threshold: RMSD threshold (A)\n",
    "        existing: a list of existing cluster IDs\n",
    "        chunksize: number of structures to put in a chunk\n",
    "          This is an implementation detail that only affects the speed, not the result\n",
    "    \n",
    "    This algorithm is used in ATTRACT and in the ProtNAff pipeline \n",
    "    (./create_frag_library/fastcluster_npy.py)\n",
    "    \"\"\"\n",
    "    if len(structures.shape) == 3:\n",
    "        assert structures.shape[2] == 3\n",
    "        structures = structures.reshape(structures.shape[0], structures.shape[1]*3)\n",
    "    if len(structures.shape) == 2:\n",
    "        assert structures.shape[1] % 3 == 0\n",
    "    natoms = structures.shape[1]/3\n",
    "    # threshold2 = sum-of-sd threshold = (RMSD threshold **2) * natoms\n",
    "    threshold2 = threshold**2 * natoms\n",
    "\n",
    "    nclus = 1\n",
    "    if len(existing):\n",
    "        clus_space = max(len(existing), 100)\n",
    "        clus = np.zeros((clus_space, structures.shape[1]))\n",
    "        clus[:len(existing)] = structures[existing]\n",
    "        clustids = []\n",
    "        clustids[:] = existing[:]\n",
    "    else:\n",
    "        clus_space = 100\n",
    "        clus = np.zeros((clus_space, structures.shape[1]))\n",
    "        clus[:1] = structures[:1]\n",
    "        clustids = [0]\n",
    "    for n in range(1, len(structures), chunksize):\n",
    "        #print(\"{0}/{1}\".format(n, len(structures), file=sys.stderr)\n",
    "        #sys.stderr.flush()\n",
    "        chunk = structures[n:n+chunksize]\n",
    "        d = chunk[:, np.newaxis, :] - clus[np.newaxis, :, :]\n",
    "        inter_sd = np.einsum(\"...ij,...ij->...i\", d, d)\n",
    "        #close_inter is a 2D Boolean matrix:\n",
    "        #  True  (1): chunk[i] is close to (within RMSD threshold of) clus[j]\n",
    "        #  False (0): chunk[i] is not close to clus[j]\n",
    "        close_inter = (inter_sd < threshold2)\n",
    "        # newclustered contains all structures in the chunk that *don't* cluster with an existing cluster\n",
    "        newclustered = []\n",
    "        for chunk_index, closest_inter in enumerate(np.argmax(close_inter,axis=1)):\n",
    "            # closest_inter contains the *first* index of close_inter\n",
    "            #   with the highest value of close_inter\n",
    "            # We are interested in the case where close_inter is all False (=> new cluster)\n",
    "            # In that case, the highest value of close_inter is False, and closest_inter is 0\n",
    "            # If close_inter is *not* all False (=> existing cluster), one of these conditions is False\n",
    "            if closest_inter == 0 and close_inter[chunk_index, 0] == False:\n",
    "                newclustered.append(chunk_index)\n",
    "\n",
    "        if len(newclustered):\n",
    "            # Now we have newclustered: the *chunk* index of all structures in the chunk that will be in new clusters\n",
    "            # Now we want to cluster them among themselves, and add the *structure* id of each new cluster\n",
    "            chunk_newclustered = chunk[newclustered]\n",
    "            d = chunk_newclustered[:, np.newaxis, :] - chunk_newclustered[np.newaxis, :, :]\n",
    "            intra_sd = np.einsum(\"...ij,...ij->...i\", d, d)\n",
    "            close_intra = (intra_sd < threshold2)\n",
    "\n",
    "            # set all upper-triangular indices to False\n",
    "            close_intra[np.triu_indices(len(chunk_newclustered))] = 0\n",
    "            for nn in range(len(chunk_newclustered)):\n",
    "                # same logic as for closest_inter;\n",
    "                #  except that we don't have the chunk index, but the chunk_newclustered index (nn)\n",
    "                #  and, since we modify close_intra in the \"else\" clause, argmax is computed later\n",
    "                closest_intra = np.argmax(close_intra[nn])\n",
    "                if closest_intra == 0 and close_intra[nn, 0] == False:\n",
    "                    chunk_index = newclustered[nn]\n",
    "                    # if clus is full, re-allocate it as a 50 % larger array\n",
    "                    if nclus == clus_space:\n",
    "                        clus_space = int(clus_space*1.5)\n",
    "                        clus_old = clus\n",
    "                        clus = np.zeros((clus_space, structures.shape[1]))\n",
    "                        clus[:nclus] = clus_old\n",
    "                    clus[nclus] = chunk[chunk_index]\n",
    "                    clustids.append(n+chunk_index)\n",
    "                    nclus += 1\n",
    "                else:  # in addition, if we aren't a new cluster, being close to us doesn't matter\n",
    "                    close_intra[:, nn] = False\n",
    "\n",
    "    return clustids\n",
    "\n",
    "def hierarchical_cluster(structures,threshold, secondary_threshold, chunksize, **kwargs):\n",
    "    \"\"\"Clusters structures using an RMSD threshold\n",
    "    Does an initial clustering using fastcluster() with a secondary threshold\n",
    "     then the resulting clusters are clustered using cluster()\n",
    "    \n",
    "    This is very close to the approach of the ProtNAff pipeline\n",
    "    (./create_frag_library/clusterfrag_npy.sh)\n",
    "    \"\"\"\n",
    "    if secondary_threshold <= threshold:\n",
    "        raise ValueError(\"Secondary threshold must be higher than clustering threshold\")\n",
    "    chunksize = 200\n",
    "    initial_clustids = fastcluster(structures,secondary_threshold,chunksize=chunksize)\n",
    "    initial_clusters = reassign(structures, initial_clustids, chunksize=chunksize)\n",
    "    all_rough_clustids = []\n",
    "    for initial_cluster in initial_clusters.values():\n",
    "        if len(initial_cluster) > 20000:\n",
    "            raise ValueError(\"Initial cluster is too large, maximum is 20 000. Increase secondary RMSD threshold.\")\n",
    "        initial_cluster = np.array(initial_cluster)\n",
    "        initial_cluster_struc = structures[initial_cluster]\n",
    "        subclustids0 = cluster(initial_cluster_struc, threshold)\n",
    "        subclustids = initial_cluster[subclustids0]\n",
    "        all_rough_clustids.append(subclustids)\n",
    "    rough_clustids = np.concatenate(all_rough_clustids)\n",
    "    rough_clust_struc = structures[rough_clustids]\n",
    "    if len(rough_clustids) > 20000:\n",
    "        clustids0 = fastcluster(rough_clust_struc, threshold, chunksize=chunksize)\n",
    "    else:\n",
    "        clustids0 = cluster(rough_clust_struc, threshold)\n",
    "    fine_clustids = rough_clustids[clustids0].tolist()\n",
    "    final_clustids = fastcluster(structures, threshold, chunksize=chunksize, existing=fine_clustids)\n",
    "    return final_clustids    \n",
    "\n",
    "clustering_methods = {\n",
    "    \"Fast\": fastcluster,\n",
    "    \"Full matrix\": cluster,\n",
    "    \"Hierarchical\": hierarchical_cluster,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "068c96f6",
   "metadata": {
    "id": "068c96f6"
   },
   "source": [
    "# User parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9c396339",
   "metadata": {
    "cellView": "form",
    "id": "9c396339"
   },
   "outputs": [],
   "source": [
    "filtering_method = \"Pure single-stranded, no NMR structures\" #@param [\"None\", \"Pure single-stranded\", \"Pure single-stranded, only NMR structures\", \"Pure single-stranded, no NMR structures\"]\n",
    "clustering_method = \"Hierarchical\" #@param [\"Fast\", \"Full matrix\", \"Hierarchical\"]\n",
    "clustering_threshold = 2.0 #@param {type:\"slider\", min:0.1, max:10, step:0.1}\n",
    "secondary_threshold = 4.0 #@param {type:\"slider\", min:0.1, max:10, step:0.1}\n",
    "chunksize = 200  # affects only the speed, not the result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70f08a50",
   "metadata": {
    "id": "70f08a50"
   },
   "source": [
    "## Parameter validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "75fc1639",
   "metadata": {
    "cellView": "form",
    "id": "75fc1639"
   },
   "outputs": [],
   "source": [
    "#@title\n",
    "if filtering_method not in filters:\n",
    "    raise ValueError(\"Filtering method '{}' not in {}\".format(filtering_method, filters.keys()))\n",
    "if clustering_method not in clustering_methods:\n",
    "    raise ValueError(\"Clustering method '{}' not in {}\".format(clustering_method, clustering_methods.keys()))\n",
    "clustering_threshold = float(clustering_threshold)\n",
    "if clustering_threshold <= 0:\n",
    "    raise ValueError(\"Clustering threshold must be positive\")\n",
    "secondary_threshold = float(secondary_threshold)\n",
    "chunksize = int(chunksize)\n",
    "if chunksize <= 0:\n",
    "    raise ValueError(\"Chunk size must be positive\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bce0230c",
   "metadata": {
    "id": "bce0230c"
   },
   "source": [
    "# Execution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edc46630",
   "metadata": {
    "id": "edc46630"
   },
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "942cf660",
   "metadata": {
    "cellView": "form",
    "id": "942cf660"
   },
   "outputs": [],
   "source": [
    "#@title\n",
    "import os\n",
    "import tempfile\n",
    "import shutil\n",
    "import json\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from urllib.request import urlretrieve\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef6a7e9f",
   "metadata": {
    "id": "ef6a7e9f"
   },
   "source": [
    "## Loading fragment library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1a125adc",
   "metadata": {
    "cellView": "form",
    "id": "1a125adc"
   },
   "outputs": [],
   "source": [
    "#@title\n",
    "fraglib_files = {\n",
    "    \"structures.json\": \"https://zenodo.org/record/6475637/files/structures.json\",\n",
    "\n",
    "    \"fragments_clust.json\": \"https://zenodo.org/record/6475637/files/fragments_clust.json\",\n",
    "\n",
    "    \"aa-fit.tgz\": \"https://zenodo.org/record/6475637/files/aa-fit.tgz\",  \n",
    "    # containing:\n",
    "    #   <motif>-aa-fit.npy\n",
    "    #   <motif>-aa.noclash\n",
    "    #   for 8 motifs (AAA to CCC)\n",
    "    # These come from the trilib/ directory.\n",
    "    # The .npy contains only the non-clashing fragments\n",
    "}\n",
    "\n",
    "if structures is None:\n",
    "    with tempfile.TemporaryDirectory() as TEMPDIR:\n",
    "        for fraglib_file, url in fraglib_files.items():\n",
    "            dest = os.path.join(TEMPDIR, fraglib_file)\n",
    "            print(\"Download\", url, \"... \", end=\"\")\n",
    "            urlretrieve(url, dest)\n",
    "            print(\"Done\")\n",
    "        os.system(\"cd {}; tar xzf aa-fit.tgz\".format(TEMPDIR))\n",
    "\n",
    "        structures = json.load(open(os.path.join(TEMPDIR, \"structures.json\")))\n",
    "        fragments = json.load(open(os.path.join(TEMPDIR, \"fragments_clust.json\")))\n",
    "        coordinates = {}\n",
    "        rev_fragindex = {}\n",
    "        for motif in motifs:\n",
    "            coorfile = \"{}-aa-fit.npy\".format(motif)\n",
    "            coordinates[motif] = np.load(os.path.join(TEMPDIR, coorfile))\n",
    "            noclash_file = \"{}-aa.noclash\".format(motif)\n",
    "            motif_rev_fragindex = {}\n",
    "            with open(os.path.join(TEMPDIR, noclash_file)) as f:\n",
    "                for lnr, l in enumerate(f.readlines()):\n",
    "                    frag_key = l.strip()\n",
    "                    motif_rev_fragindex[frag_key] = lnr\n",
    "                assert len(motif_rev_fragindex) == len(coordinates[motif]), (len(motif_rev_fragindex), noclash_file, len(coordinates[motif]), coorfile)\n",
    "                rev_fragindex[motif] = motif_rev_fragindex\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00222529",
   "metadata": {
    "id": "00222529"
   },
   "source": [
    "## Run filtering and clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "412c4df6",
   "metadata": {
    "cellView": "form",
    "id": "412c4df6"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Motif AAA, filtering... clustering... Done\n",
      "Motif AAC, filtering... clustering... Done\n",
      "Motif ACA, filtering... clustering... Done\n",
      "Motif ACC, filtering... clustering... Done\n",
      "Motif CAA, filtering... clustering... Done\n",
      "Motif CAC, filtering... clustering... Done\n",
      "Motif CCA, filtering... clustering... Done\n",
      "Motif CCC, filtering... clustering... Done\n"
     ]
    }
   ],
   "source": [
    "#@title\n",
    "filt = filters[filtering_method]\n",
    "clustering_func = clustering_methods[clustering_method]\n",
    "filtered_coordinates = filt()\n",
    "clusters = {}\n",
    "for motif in motifs:\n",
    "    print(\"Motif {}, filtering... \".format(motif),end=\"\")\n",
    "    motif_coors = filtered_coordinates[motif]\n",
    "    print(\"clustering... \", end=\"\")\n",
    "    motif_clustids = clustering_func(\n",
    "        motif_coors,\n",
    "        threshold=clustering_threshold,\n",
    "        secondary_threshold=secondary_threshold,\n",
    "        chunksize=chunksize\n",
    "    )\n",
    "    motif_clusters = reassign(\n",
    "        motif_coors,\n",
    "        motif_clustids,\n",
    "        chunksize=chunksize\n",
    "    )\n",
    "    print(\"Done\")\n",
    "    clusters[motif] = motif_clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4bf44c1",
   "metadata": {
    "id": "d4bf44c1"
   },
   "source": [
    "# Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9513e5a7",
   "metadata": {
    "id": "9513e5a7"
   },
   "source": [
    "## Number of clusters per motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "574cf777",
   "metadata": {
    "cellView": "form",
    "id": "574cf777",
    "outputId": "5a9a878d-50ac-44a7-a977-1ad5db775b7f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAA 857\n",
      "AAC 675\n",
      "ACA 580\n",
      "ACC 641\n",
      "CAA 649\n",
      "CAC 435\n",
      "CCA 669\n",
      "CCC 1003\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAmQAAAE1CAYAAABJK/xzAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAgUklEQVR4nO3df7RcZX3v8feXEAyGEH5ZESIkGEQwpECPKKUKopagJgKiJSirYCR2WZH+0AqKIsWKCtaKghorgtIFplx/EJsLqShJ9XoLBAETA02McjnBNhBKQBT5ke/9Y+bQyck5c/Y5mT37zMz7tdasM/vZs2c++2EyfNfez352ZCaSJEmqzg5VB5AkSep1FmSSJEkVsyCTJEmqmAWZJElSxSzIJEmSKmZBJkmSVDELMkmSpIpZkEmSJFVsXBVkETE5IlZGxBurziJJktQupRZkEXFlRGyMiFWD2udExL0RsS4izm1Y9QFgcZmZJEmSxpso89ZJEfEq4NfA1zJzVr1tAvAfwOuAfuA2YD6wD7AXMAl4KDO/W1owSZKkcWTHMt88M1dExPRBzUcC6zJzPUBEXAe8CdgFmAwcAvw2IpZm5pYy80mSJI0HpRZkw9gXuL9huR94eWa+ByAizqB2hGzIYiwiFgILASZPnvwHL3nJS8pNK0mS1AIrV658KDOfN9S6KgqyGKLt2fOmmXlVs40zcxGwCKCvry9vv/32loaTJEkqQ0TcN9y6Kq6y7Ade2LA8DXigghySJEnjQhUF2W3AgRExIyJ2Ak4FbhjNG0TE3IhYtHnz5lICSpIktVPZ015cC/wYOCgi+iNiQWY+DbwHuAlYAyzOzNWjed/MXJKZC6dOndr60JIkSW1W9lWW84dpXwosLfOzJUlSaz311FP09/fzxBNPVB1lXJs0aRLTpk1j4sSJhbepYlC/JEnqQP39/UyZMoXp06cTMdQ1espMNm3aRH9/PzNmzCi83bi6dVJRjiGTJKn9nnjiCfbcc0+LsSYigj333HPURxE7siBzDJkkSdWwGBvZWPqoIwsySZIkgI9+9KNceumlo97ukUce4Yorrigh0dg4hkySJI1JXNjao2V5QXn31x5soCB797vfXXibzCQz2WGH1h/P6sgjZI4hkySpN33ta19j9uzZ/P7v/z6nn376VuuOPfZYBu7g89BDDzF9+nQAVq9ezZFHHslhhx3G7NmzWbt2Leeeey4///nPOeyww3j/+98PwCWXXMLLXvYyZs+ezQUXXADAL3/5Sw4++GDe/e53c8QRR3D//fdzxhlnMGvWLA499FA+85nPtGS/OvIIWWYuAZb09fWdVXUWSZLUHqtXr+bv/u7v+NGPfsRee+3Fww8/zGWXXTbidl/84hc555xzeNvb3saTTz7JM888wyc+8QlWrVrFnXfeCcCyZctYu3Ytt956K5nJvHnzWLFiBfvttx/33nsvX/3qV7niiitYuXIlGzZsYNWqVUDtSFsrdOQRMkmS1Hu+//3vc8opp7DXXnsBsMceexTa7qijjuLjH/84n/zkJ7nvvvvYeeedt3nNsmXLWLZsGYcffjhHHHEE99xzD2vXrgVg//335xWveAUABxxwAOvXr+fss8/mxhtvZNddd23JvlmQSZKkjpCZTa9g3HHHHdmyZQvAVtNOnHbaadxwww3svPPOHH/88Xz/+98f8r3PO+887rzzTu68807WrVvHggULAJg8efKzr9t999256667OPbYY7n88st55zvf2ZJ968iCzDFkkiT1nte85jUsXryYTZs2AfDwww9vtX769OmsXLkSgOuvv/7Z9vXr13PAAQfw3ve+l3nz5nH33XczZcoUHnvssWdfc/zxx3PllVfy61//GoANGzawcePGbTI89NBDbNmyhTe/+c1cdNFF3HHHHS3ZN8eQSZKkjvDSl76UD33oQxxzzDFMmDCBww8//NmB+wDve9/7eOtb38rXv/51jjvuuGfbv/GNb3DNNdcwceJE9t57bz7ykY+wxx57cPTRRzNr1ixOOOEELrnkEtasWcNRRx0FwC677MI111zDhAkTtsqwYcMGzjzzzGePxF188cUt2bfIbN8lpq3W19eXA1dTSJKkcq1Zs4aDDz646hgdYai+ioiVmdk31Os78pSlJElSN7EgkyRJqlhHFmQO6pckSd2kIwsyby4uSVI1OnnsebuMpY86siCTJEntN2nSJDZt2mRR1kRmsmnTJiZNmjSq7Tpy2gtJktR+06ZNo7+/nwcffLDqKOPapEmTmDZt2qi2sSCTJEmFTJw4kRkzZlQdoyt5ylKSJKliFmSSJEkV68iCzGkvJElSN+nIgsxpLyRJUjfpyIJMkiSpm1iQSZIkVcyCTJIkqWIWZJIkSRWzIJMkSaqYBZkkSVLFLMgkSZIq1pEFmRPDSpKkbtKRBZkTw0qSpG7SkQWZJElSN7EgkyRJqpgFmSRJUsUsyCRJkipmQSZJklQxCzJJkqSKWZBJkiRVzIJMkiSpYjuO9IKI6ANeCewD/BZYBXwvMx8uOZskSVJPGPYIWUScERF3AOcBOwP3AhuBPwL+NSKujoj92hNzm2zeOkmSJHWNZkfIJgNHZ+Zvh1oZEYcBBwL/r4RcTWXmEmBJX1/fWe3+bEmSpFYbtiDLzMubbZiZd7Y8jSRJUg8acVB/RLw4Im6OiFX15dkRcX750SRJknpDkassv0xtHNlTAJl5N3BqmaEkSZJ6SZGC7LmZeeugtqfLCCNJktSLihRkD0XEi4AEiIhTgF+VmkqSJKmHjDgPGfDnwCLgJRGxAfgF8PZSU0mSJPWQEQuyzFwPvDYiJgM7ZOZj5ceSJEnqHcMWZBHxV8O0A5CZf19SJkmSpJ7S7AjZlLalkCRJ6mHNJoa9sJ1BJEmSelWRm4tPAhYALwUmDbRn5jtKzCVJktQzikx78XVgb+B4YDkwDXBgvyRJUosUKchmZuaHgccz82rgDcCh5caSJEnqHUUKsqfqfx+JiFnAVGB6aYkkSZJ6TJGJYRdFxO7A+cANwC7AR0pNJUmS1EOKTAz7j/WnK4ADygoSEQcD5wB7ATdn5hfK+ixJkqTxZMRTlhHx8YjYrWF594j4WJE3j4grI2JjRKwa1D4nIu6NiHURcS5AZq7JzD8D3gr0jWovJEmSOliRMWQnZOYjAwuZ+d/A6wu+/1XAnMaGiJgAXA6cABwCzI+IQ+rr5gE/BG4u+P6SJEkdr0hBNiEinjOwEBE7A89p8vpnZeYK4OFBzUcC6zJzfWY+CVwHvKn++hsy8w+BtxV5f0mSpG5QZFD/NcDNEfFVIIF3AFdvx2fuC9zfsNwPvDwijgVOplbsLR1u44hYCCwE2G+//bYjhiRJGq/iwmjr5+UF2dbPG6zIoP5PRcTdwGuBAC7KzJu24zOH6uHMzFuAWwrkWQQsAujr66u29yRJklqgyK2TJgPLMvPGiDgIOCgiJmbmUyNtO4x+4IUNy9OAB8b4XpIkSR2vyBiyFcCkiNgX+B5wJrXB+mN1G3BgRMyIiJ2AU6nNb1ZYRMyNiEWbN2/ejhiSJEnjQ5GCLDLzN9TGd30uM0+idnXkyBtGXAv8mNpRtf6IWJCZTwPvAW4C1gCLM3P1aEJn5pLMXDh16tTRbCZJkjQuFRnUHxFxFLUrHxeMYjsyc/4w7UtpMnB/POm1QYWSJKn9ihwhOwc4D/hWZq6OiAOAH5QbS5IkqXcUucpyBbVxZAPL64H3lhlqJBExF5g7c+bMKmNIkiS1RJEjZOOOY8gkSVI36ciCTJIkqZt0ZEHmtBeSJKmbFJkYdgZwNjC98fWZOa+8WM1l5hJgSV9f31lVZZAkSWqVItNXfBv4CrAE2FJqGnWEdk4F4jQgknqFv629rUhB9kRmXlZ6Ekldw/+xbMs+kdRMkYLssxFxAbAM+N1AY2beUVoqSZKkHlKkIDsUOB04jv85ZZn15Uo4D5kkSeomRQqyk4ADMvPJssMU5aB+jTeejpIkbY8i017cBexWcg5JkqSeVeQI2fOBeyLiNrYeQ1bZtBeSJEndpEhBdkHpKSRJknpYkZuLL29HkNFwUL8kSeomI44hi4iTI2JtRGyOiEcj4rGIeLQd4YbjzcUlSVI3KXLK8lPA3MxcU3YYSZKkXlTkKsv/shiTJEkqz7BHyCLi5PrT2yPiG9Tuadl4leU3y40mSZLUG5qdspzb8Pw3wB83LCdgQSZJGpN2TqYMTqis8W/YgiwzzwSIiKMz80eN6yLi6LKDSZIk9YoiY8g+V7CtbSJibkQs2rx5c5UxJEmSWqLZGLKjgD8EnhcRf9WwaldgQtnBmvFelpIkqZs0G0O2E7BL/TVTGtofBU4pM5QkSVIvaTaGbDmwPCKuysz72phJkiSppzQ7ZfkPmfkXwOcjYpvLU7y5uCRJUms0O2X59frfS9sRRJIkqVc1O2W5MiImAGdl5tvbmEmSJKmnNJ32IjOfoXaV5U5tyiNJktRzitxc/JfAjyLiBuDxgcbM/PuyQo0kIuYCc2fOnFlVBEmSpJYpMjHsA8B366+d0vCoTGYuycyFU6dOrTKGJElSS4x4hCwzL2xHEEmSpF41YkEWEc8D/gZ4KTBpoD0zjysxlyRJUs8ocsryn4B7gBnAhdTGlN1WYiZJkqSeUqQg2zMzvwI8lZnLM/MdwCtKziVJktQzilxl+VT9768i4g3UBvlPKy+SJElSbylSkH0sIqYCfw18DtgV+MtSU0mSJPWQIldZfrf+dDPw6nLjSJIk9Z4iV1leNkTzZuD2zPxO6yNJkiT1liKD+icBhwFr64/ZwB7Agoj4h9KSSZIk9YgiY8hmAsdl5tMAEfEFYBnwOuCnJWaTJEnqCUWOkO0LTG5YngzsU7/x+O9KSTWCiJgbEYs2b95cxcdLkiS1VJGC7FPAnRHx1Yi4CvgJcGlETAa+V2a44XgvS0mS1E2KXGX5lYhYChwJBPDBzHygvvr9ZYaTJEnqBUXGkAG8DHhl/fkz1CaHlSRJUguMeMoyIj4BnAP8rP54b0RcXHYwSZKkXlHkCNnrgcMycwtARFxNbRzZeWUGkyRJ6hVFBvUD7Nbw3JH0kiRJLVTkCNnFwE8i4gfUBvW/Co+OSZIktUyRqyyvjYhbqA3sD+ADmfmfZQeTJEnqFcMWZBFxxKCm/vrffSJin8y8o7xYkiRJvaPZEbJPN1mXwHEtziJJktSThi3IMvPV7QwiSZLUq4a9yjIi/qjZhhGxa0TMan0kSZKk3tLslOWbI+JTwI3ASuBBYBIwE3g1sD/w16UnlCRJ6nLNTln+ZUTsDpwCvAV4AfBbYA3wpcz8YXsiSpIkdbem015k5n8DX64/JEmSVIKiM/VLkiSpJOOqIIuIEyPiyxHxnYj446rzSJIktUPpBVlEXBkRGyNi1aD2ORFxb0Ssi4hzATLz25l5FnAG8CdlZ5MkSRoPRizIIuItETGl/vz8iPjmELP4N3MVMGfQe04ALgdOAA4B5kfEIQ0vOb++XpIkqesVOUL24cx8rD4v2fHA1cAXin5AZq4AHh7UfCSwLjPXZ+aTwHXAm6Lmk8D/9tZMkiSpVxQpyJ6p/30D8IXM/A6w03Z+7r7A/Q3L/fW2s4HXAqdExJ8NtWFELIyI2yPi9gcffHA7Y0iSJFWv6bQXdRsi4kvUCqVPRsRz2P6xZzFEW2bmZcBlzTbMzEXAIoC+vr7czhySJEmVK1JYvRW4CZiTmY8AewDv387P7Qde2LA8DXhgO99TkiSpIzU9QhYROwC3Zuaz96zMzF8Bv9rOz70NODAiZgAbgFOB04puHBFzgbkzZ87czhiSJEnVa3qELDO3AHdFxH5j/YCIuBb4MXBQRPRHxILMfBp4D7Ujb2uAxZm5uuh7ZuaSzFw4derUscaSJEkaN4qMIXsBsDoibgUeH2jMzHlFPiAz5w/TvhRYWuQ9JEmSulmRguzC0lOMkqcsJUlSNxlxUH9mLgd+CUysP78NqHSOME9ZSpKkblJkpv6zgOuBL9Wb9gW+XWImSZKknlJk2os/B44GHgXIzLXA75UZSpIkqZcUKch+V7+9EQARsSNQ6YSsETE3IhZt3ry5yhiSJEktUaQgWx4RHwR2jojXAf8MLCk3VnOOIZMkSd2kSEF2LvAg8FPgXcDSzPxQqakkSZJ6SJFpL87OzM8CXx5oiIhz6m2SJEnaTkWOkP3pEG1ntDiHJElSzxr2CFlEzKd2f8kZEXFDw6pdgU1lB2vGiWElSVI3aXbK8v9Qu4n4XsCnG9ofA+4uM9RIMnMJsKSvr++sKnNIkiS1wrAFWWbeB9wXEa8FfpuZWyLixcBLqA3wlyRJUgsUGUO2ApgUEfsCNwNnAleVGUqSJKmXFCnIIjN/A5wMfC4zTwIOKTeWJElS7yhUkEXEUcDbgH+ptxWZLqM0ztQvSZK6SZGC7C+A84BvZebqiDgA+EGpqUbgTP2SJKmbjHikKzOXA8sbltcD7y0zlCRJUi8ZsSCLiB8wxM3EM/O4UhJJkiT1mCJjwd7X8HwS8Gbg6XLiSJIk9Z4ipyxXDmr6UUQsH/LFkiRJGrUipyz3aFjcAfgDYO/SEkmSJPWYIqcsV1IbQxbUTlX+AlhQZqiReC9LSZLUTYqcspzRjiCj4b0sJUlSNxm2IIuIk5ttmJnfbH0cSZKk3tPsCNncJusSsCCTJElqgWELssw8s51BJEmSetWIt06KiI9HxG4Ny7tHxMdKTSVJktRDitzL8oTMfGRgITP/G3h9aYkkSZJ6TJGCbEJEPGdgISJ2Bp7T5PWSJEkahSLzkF0D3BwRX6U2mP8dwNWlphqB85BJkqRuMuIRssz8FPAx4GDgpcBF9bbKZOaSzFw4derUKmNIkiS1RJEjZGTmjcCNJWeRJEnqSUXGkEmSJKlEFmSSJEkVG7Ygi4ib638/2b44kiRJvafZGLIXRMQxwLyIuA6IxpWZeUepySRJknpEs4LsI8C5wDTg7wetS+C4skJJkiT1kmb3srweuD4iPpyZF7UxkyRJUk8ZcdqLzLwoIuYBr6o33ZKZ3y03liRJUu8ocnPxi4FzgJ/VH+fU2yRJktQCRSaGfQNwWGZuAYiIq4GfAOeVGUySJKlXFJ2HbLeG55Xfrygi5kbEos2bN1cdRZIkabsVKcguBn4SEVfVj46tBD5ebqzmvJelJEnqJkUG9V8bEbcAL6M2F9kHMvM/yw4mSZLUK4reXPxXwA0lZ5EkSepJ3stSkiSpYhZkkiRJFWtakEXEDhGxql1hJEmSelHTgqw+99hdEbFfm/JIkiT1nCKD+l8ArI6IW4HHBxozc15pqSRJknpIkYLswtJTSJIk9bAi85Atj4j9gQMz83sR8VxgQvnRJEmSekORm4ufBVwPfKnetC/w7RIzSZIk9ZQi0178OXA08ChAZq4Ffq/MUJIkSb2kSEH2u8x8cmAhInYEsrxIkiRJvaVIQbY8Ij4I7BwRrwP+GVhSbixJkqTeUaQgOxd4EPgp8C5gKXB+maEkSZJ6SZGrLLdExNXAv1M7VXlvZnrKUpIkqUWKXGX5BuDnwGXA54F1EXFCq4NExAER8ZWIuL7V7y1JkjSeFTll+Wng1Zl5bGYeA7wa+EyRN4+IKyNi4+D7YUbEnIi4NyLWRcS5AJm5PjMXjHYHJEmSOl2RgmxjZq5rWF4PbCz4/lcBcxobImICcDlwAnAIMD8iDin4fpIkSV1n2DFkEXFy/enqiFgKLKY2huwtwG1F3jwzV0TE9EHNRwLrMnN9/XOuA94E/Gx00SVJkrpDsyNkc+uPScB/AccAx1K74nL37fjMfYH7G5b7gX0jYs+I+CJweEScN9zGEbEwIm6PiNsffPDB7YghSZI0Pgx7hCwzzyzpM2Poj8tNwJ+NtHFmLgIWAfT19Xm1pyRJ6ngjTnsRETOAs4Hpja/PzHlj/Mx+4IUNy9OAB8b4XpIkSR1vxIKM2o3Ev0Jtdv4tLfjM24AD64XeBuBU4LTRvEFEzAXmzpw5swVxJEmSqlXkKssnMvOyzPxBZi4feBR584i4FvgxcFBE9EfEgsx8GngPcBOwBlicmatHEzozl2TmwqlTp45mM0mSpHGpyBGyz0bEBcAy4HcDjZl5x0gbZub8YdqXUrsFkyRJUs8rUpAdCpwOHMf/nLLM+nIlPGUpSZK6SZGC7CTggMx8suwwRWXmEmBJX1/fWVVnkSRJ2l5FxpDdBexWcg5JkqSeVeQI2fOBeyLiNrYeQzbWaS8kSZLUoEhBdkHpKUbJMWSSJKmbjFiQFZ3iop0cQyZJkrpJkZn6H6N2VSXATsBE4PHM3LXMYJIkSb2iyBGyKY3LEXEicGRZgSRJknpNkasst5KZ36bCOcgkSZK6TZFTlic3LO4A9PE/pzAr4aB+SZLUTYpcZTm34fnTwC+BN5WSpiAH9UuSpG5SZAzZme0IIkmS1KuGLcgi4iNNtsvMvKiEPJIkST2n2RGyx4domwwsAPYELMgkSZJaYNiCLDM/PfA8IqYA5wBnAtcBnx5uu3ZwUL8kSeomTae9iIg9IuJjwN3UircjMvMDmbmxLemGkZlLMnPh1KlTq4whSZLUEs3GkF0CnAwsAg7NzF+3LZUkSVIPaXaE7K+BfYDzgQci4tH647GIeLQ98SRJkrpfszFko57FX5IkSaNn0SVJklQxCzJJkqSKdWRBFhFzI2LR5s2bq44iSZK03TqyIHPaC0mS1E06siCTJEnqJhZkkiRJFbMgkyRJqpgFmSRJUsUsyCRJkipmQSZJklSxjizInIdMkiR1k44syJyHTJIkdZOOLMgkSZK6iQWZJElSxSzIJEmSKmZBJkmSVDELMkmSpIpZkEmSJFXMgkySJKliFmSSJEkVsyCTJEmqmAWZJElSxTqyIPNelpIkqZt0ZEHmvSwlSVI36ciCTJIkqZtYkEmSJFXMgkySJKliFmSSJEkVsyCTJEmqmAWZJElSxSzIJEmSKmZBJkmSVDELMkmSpIpZkEmSJFXMgkySJKliFmSSJEkVsyCTJEmqmAWZJElSxSzIJEmSKrZj1QEGRMRk4ArgSeCWzPyniiNJkiS1RalHyCLiyojYGBGrBrXPiYh7I2JdRJxbbz4ZuD4zzwLmlZlLkiRpPCn7lOVVwJzGhoiYAFwOnAAcAsyPiEOAacD99Zc9U3IuSZKkcaPUgiwzVwAPD2o+EliXmesz80ngOuBNQD+1oqz0XJIkSeNJZGa5HxAxHfhuZs6qL58CzMnMd9aXTwdeDnwA+DzwBPDD4caQRcRCYGF98SDg3lJ3YOz2Ah6qOsQ4Y59szf7Yln2yNftjW/bJ1uyPbY3nPtk/M5831IoqBvXHEG2ZmY8DZ460cWYuAha1PFWLRcTtmdlXdY7xxD7Zmv2xLftka/bHtuyTrdkf2+rUPqni1GA/8MKG5WnAAxXkkCRJGheqKMhuAw6MiBkRsRNwKnBDBTkkSZLGhbKnvbgW+DFwUET0R8SCzHwaeA9wE7AGWJyZq8vMUZFxf1q1AvbJ1uyPbdknW7M/tmWfbM3+2FZH9knpg/olSZLUnNNLSJIkVcyCbIwi4qSIyIh4yaD2w+vtxw+xzfMi4qmIeFf7krbHaPsjIl4cEUvrd2tYExGLI+L57U1drqH6JCKOjIgV9TtV3BMR/xgRz21Y/52I+HE1ics32j6JiBMi4vb6d+SeiLi0uvStFRF7R8R1EfHziPhZ/d/Di+vr/jIinoiIqUNsN+S/tW4wlj7p5u8IDN8nI/2GRsRnI2JDRHTV/+fH0h8j/e6OF131H6rN5gM/pHZRwlDt84fY5i3A/x1mXacr3B8RMQn4F+ALmTkzMw8GvgAMOTdLB9uqT+o/Dv8MfCAzDwIOBm4EptTX7wYcAewWETOqCNwGhfskImZRm5vw7fXvyCxgfSWpWywiAvgWtfv2vigzDwE+CAz8D3U+tQugThpi8+H+rXW0sfRJN39HYMQ+GfY3tF6EnUTt7jevqiR8CcbSHyP97o4rmeljlA9gF2AD8GLgnob2oPZj8CJqU3lMGrTdv1G/UwGwb9X7UVV/AO8AvlZ17nb3CfC3wN822WYBcAVwAXBe1ftQdZ8AXwPeUXXukvriOGDFMOteBPwMOAa4aaQ+7JbHWPqkm78jzfpkpN9Q4DXAUuBPgS9VvR9V9sdIv7vj6eERsrE5EbgxM/8DeDgijqi3Hw38IjN/DtwCvH5gg4h4IbB3Zt4KLAb+pK2Jy3Uio+uPWcDKdodssxPZtk9G2u/5wLX1RzceRT2R0fVJN39Pmu3bwPfg36hdof57DetOZOh/a91gLH3Szd8RGH7/iv6WfAt4Y0RMLCFbFcbSHx3zHbEgG5v51O7BSf3v/BHaoXZ6YfEw6zrdWPqj241q3+uH1WdSu23YfwBP10/HdJNe/j6MxqnAdZm5BfgmtaEOA3q1D5v1iRpEbX7P1wPfzsxHgX8H/rjaVCqiilsndbSI2JPaYdNZEZHABCAj4lzgzcC8iPgQtdN1e0bElMx8jNoP5/Mj4m31t9onIg7MzLUV7EbLjKU/gNXUTj10peH6BLga+APgO0Ns9ifA7sAvasMk2JXa/4TOb0fmso2xT1bX193VrpxttBo4ZXBjRMwGDgT+tf492Inaaf/Lm/xb+5usn5vpcKPuE7r7OwLD9AnNf0PnAFOBn9b767nAb6iNsep0Y+mPge/IUL8x40vV50w77QG8i0Hn5IHl1Mb9DB7vcTVwOvWboA9adyHw4ar3p6L+2JnaOLo3NKybAxxa9f6U3CfHAPcBL29ofzuwN7UJlI9qaJ8BrKt6Xyruk9n178mL6+07AH9V9b60qD+C2pGLsxraXgb8gEHjB4FfAPs36cNXVr0/FfZJ135HRuiTY4b7DaU+5KGhfTKwEXhu1ftTUX88f7jfmKr3Z/DDU5ajN5/aeflG/wt4xTDtpzXZphtON4y6PzLzt8AbgbMjYm1E/Aw4g9qPRjcYrk9OrT8urV9+vQZ4JbAHsB+1K3AByMxfAI9GxMvbE7l0o+2TRzPzbuAvgGvr7auAF7Qvcnmy9n+Fk4DX1S/fXw18FDiWbfvpW9T6aLg+PK3UsG0ylj7p5u8INO2TBxj6N/RR4HgajoZl5uPUrsqd2970rTeG/tiYmf/FML8xVexDM87UL0mSVDGPkEmSJFXMgkySJKliFmSSJEkVsyCTJEmqmAWZJElSxSzIJEmSKmZBJkmSVDELMkmSpIr9f3ncxOedXpztAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#@title\n",
    "from math import log10, ceil\n",
    "\n",
    "nclusts = []\n",
    "for motif in motifs:\n",
    "    nclust = len(clusters[motif])\n",
    "    print(motif, nclust)\n",
    "    nclusts.append(nclust)\n",
    "plt.figure(figsize=(10,5))\n",
    "plt.bar([1, 2, 3, 4, 5, 6, 7, 8], nclusts, 0.4, label=\"clusters\", log=True, color='green')\n",
    "plt.ylabel(\"Number of clusters (logarithm scale)\")\n",
    "plt.xticks([1, 2, 3, 4, 5, 6, 7, 8], motifs)\n",
    "ylim = 10**(ceil(log10(max(nclusts)+1)))\n",
    "plt.ylim(1, ylim)\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41e29c37",
   "metadata": {
    "id": "41e29c37"
   },
   "source": [
    "## Distribution of cluster sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6790d77a",
   "metadata": {
    "cellView": "form",
    "id": "6790d77a",
    "outputId": "56441128-20bb-487c-e4f9-326211852d74"
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfUAAAEKCAYAAAALjMzdAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAg1klEQVR4nO3dfZRdZ13o8e8v06FOETKUFmynDS2khlupEhnK5UV5sZgipMQUhMryCnYRXgTrXYt4Gy8KXri2vbEqXBAIiMW1sKXUGFsoDVr64gWUJoQ2DSW2VrGZoK1AwktHmia/+8fZU85Mzss+M2fPeZnvZ61ZmfOcvZ/nd/bszG/2fp79PJGZSJKkwbes1wFIkqTuMKlLkjQkTOqSJA0Jk7okSUPCpC5J0pAwqUuSNCSO6XUAC3XCCSfkaaed1uswJElaFDt37vyPzDyx0XsDn9RPO+00duzY0eswJElaFBHx9WbveftdkqQhYVKXJGlIDGxSj4i1EbHl4MGDvQ5FkqS+MLBJPTOvy8wNy5cv73UokiT1hYEfKNct23ZNsXn7XvYfmObk8TE2rlnFutUTvQ5LkqTSTOrUEvqmrbuZPnQYgKkD02zauhvAxC5JGhgDe/u9mzZv3/tIQp8xfegwm7fv7VFEkiR1zqQO7D8w3VG5JEn9yKQOnDw+1lG5JEn9aGCTejcfadu4ZhVjoyOzysZGR9i4ZtWC65YkabEMbFLv5iNt61ZPcMn6s5gYHyOAifExLll/loPkJEkDxdHvhXWrJ0zikqSBNrBX6pIkaTaTuiRJQ8KkLknSkDCpS5I0JEzqkiQNCZO6JElDwqQuSdKQGNik3s0Z5SRJGgYDm9S7OaOcJEnDYGCTuiRJms2kLknSkDCpS5I0JEzqkiQNCZO6JElDwqQuSdKQMKlLkjQkTOqSJA0Jk7okSUPimF4HoOps2zXF5u172X9gmpPHx9i4ZhXrVk/0OixJUkVM6kNq264pNm3dzfShwwBMHZhm09bdACZ2SRpSfXX7PSL+S0R8MCKuiYg39TqeQbZ5+95HEvqM6UOH2bx9b48ikiRVrfKkHhEfjYj7I+LOOeXnRsTeiLgnIi4GyMy7MvONwC8Bk1XHNsz2H5juqFySNPgW40r9CuDc+oKIGAHeD7wEOBO4ICLOLN47D/h/wI2LENvQOnl8rKNySdLgqzypZ+atwLfmFJ8N3JOZ92bmQ8BVwMuL7a/NzOcAr6k6tmG2cc0qxkZHZpWNjY6wcc2qHkUkSaparwbKTQD31b3eBzwrIl4ArAeOBa5vtnNEbAA2AKxYsaKyIAfZzGA4R79L0tLRq6QeDcoyM28Gbm63c2ZuAbYATE5OZlcjGyLrVk+YxCVpCenV6Pd9wKl1r08B9ndSQUSsjYgtBw8e7GpgkiQNql4l9duAMyLi9Ih4FPBq4NpOKsjM6zJzw/LlyysJUJKkQbMYj7RdCXwRWBUR+yLiwsx8GHgLsB24C7g6M/dUHYskScOs8j71zLygSfn1tBgM105ErAXWrly5cr5VSJI0VPpqRrlOePtdkqTZBjapS5Kk2QY2qTv6XZKk2QY2qXv7XZKk2QY2qUuSpNlM6pIkDYmBTer2qUuSNNvAJnX71CVJmm1gk7okSZrNpC5J0pAwqUuSNCQGNqk7UE6SpNkGNqk7UK6/bds1xXMv/RynX/xpnnvp59i2a6rXIUnS0Kt8lTYtPdt2TbFp626mDx0GYOrANJu27gZg3eqJXoYmSUNtYK/U1b82b9/7SEKfMX3oMJu37+1RRJK0NJjU1XX7D0x3VC5J6o6BTeoOlOtfJ4+PdVQuSeqOgU3qDpTrXxvXrGJsdGRW2djoCBvXrOpRRJK0NDhQTl03Mxhu8/a97D8wzcnjY2xcs8pBcpJUsbZJPSJeCdyQmd+NiLcDPw28OzO/XHl0GljrVk+YxCVpkZW5/f47RUJ/HrAG+BjwgWrDkiRJnSpz+33m2aSXAh/IzL+OiHdWF5LU3LZdU97Wl6QmyiT1qYj4EHAOcFlEHMsAD7DT4HJSG0lqrUxy/iVgO3BuZh4Ajgc2VhlUGT7StvQ4qY0ktdYyqUfEMuBLmbk1M+8GyMxvZOZnFyW6FnykbelxUhtJaq1lUs/MI8DtEbFikeKRmnJSG0lqrczt95OAPRFxY0RcO/NVdWDSXE5qI0mtlRko93uVRyGV4KQ2ktRa26SembdExJOAMzLzbyPiOGCk3X5SFZzURpKaa3v7PSJeD1wDfKgomgC2VRiTJEmahzJ96r8OPBf4DkAxCv4JVQYlSZI6Vyap/yAzH5p5ERHHAFldSJIkaT7KJPVbIuK3gbGIeDHwSeC6asOSJEmdKpPULwYeAHYDbwCuz8z/WWlUJTijnCRJs5VJ6m/NzA9n5isz8xWZ+eGIuKjyyNpwRjlJkmYrk9R/tUHZa7schyRJWqCmz6lHxAXALwOnz5lB7rHAN6sOTJIkdabV5DNfAL4BnABcXlf+XeCOKoOSJEmda5rUM/PrwNcj4hxgOjOPRMSPA0+lNmhOkiT1kTJzv98K/ExEPA64EdgBvAp4TZWBSVXZtmvK+eMlDaUyA+UiMx8E1gP/NzN/ETiz2rCkamzbNcWmrbuZOjBNAlMHptm0dTfbdk31OjRJWrBSST0ink3tyvzTRVmZK3yp72zevpfpQ4dnlU0fOszm7Xt7FJEkdU+ZpP6bwCbgrzJzT0Q8Gbip0qikiuw/MN1RuSQNklJLrwK31L2+F/iNKoOSqnLy+BhTDRL4yeNjPYhGkrqrbVKPiJtosIBLZr6okoikCm1cs4pNW3fPugU/NjrCxjWrehiVJHVHmb7xt9V9/yPA+cDD1YQjVWtmlLuj3yUNozK333fOKfp8RNzScGNpAKxbPWESlzSUytx+P77u5TLgGcCPVRFMRKwDXgo8AXh/Zn62inYkSRpGZW6/76TWpx7Ubrv/M3Bh2QYi4qPAy4D7M/NpdeXnAu8BRoCPZOalmbkN2FZMdPMHgEldkqSSytx+P32BbVwBvA/485mCiBgB3g+8GNgH3BYR12bmV4tN3l68L/UNZ6KT1O9ardK2vtWOmbm1TAOZeWtEnDan+GzgnuLxOCLiKuDlEXEXcCnwmcz8covYNgAbAFasWFEmDGlBZmaimxk1PzMTHWBil9Q3Wl2pr23xXgKlknoTE8B9da/3Ac8C3gqcAyyPiJWZ+cGGjWduAbYATE5OHvW4ndRtrWaiM6lL6hetVml7XYXtRuMm873AeytsV5oXZ6KTNAjaThMbEb8fEeN1rx8XEe9eYLv7gFPrXp8C7O+kgohYGxFbDh48uMBQpPaazTjnTHSS+kmZud9fkpkHZl5k5reBX1hgu7cBZ0TE6RHxKODVwLWdVJCZ12XmhuXLly8wFKm9jWtWMTY6MqvMmegk9ZsySX0kIo6deRERY8CxLbafJSKuBL4IrIqIfRFxYWY+DLwF2A7cBVydmXs6C11aPOtWT3DJ+rOYGB8jgInxMS5Zf5b96ZL6SmS2HmcWEb8FnAf8GbUBcr8GXJuZ/6f68FrGtRZYu3LlytfffffdvQxFkqRFExE7M3Oy4XvtknpRwbnURqUH8NnM3N7dEOdvcnIyd+zY0eswJElaFK2SepkZ5cjMG4AbuhqVJEnqqjJ96n3J0e+SJM02sEnd0e+SJM3WNKlHxI3Fv5ctXjiSJGm+WvWpnxQRzwfOK+ZmnzULXKu52SVJ0uJrldR/F7iY2mxvfzjnvQReVFVQZdQ90tbLMKQFc/U3Sd1S5jn138nMdy1SPB3zkTYNsrmrv0FtpjontpHUTKtH2toOlMvMd0XEeRHxB8XXy7oforQ0tVr9TZI6VWZBl0uAi4CvFl8XFWWSFsjV3yR1U5nJZ14KPD0zjwBExMeAXcCmKgNrxz51DYOTx8eYapDAXf1N0nyUfU59vO77vngw3OfUNQxc/U1SN5W5Ur8E2BURN1F7rO1n6fFVujQsZgbDOfpdUjeUXdDlJOCZ1JL6P2Tmv1UdWFmOfpckLSXdWNDlG8C1XY1KGmI+ey6pF0oldUnlzX32fOrANJu27gYwsUuq1MAu6OIqbepXPnsuqVdaJvWIWBYRdy5WMJ1w9Lv6lc+eS+qVlrffM/NIRNweESsy818XKyhpkFXx7Ll99JLKKHP7/SRgT0TcGBHXznxVHZg0qLr97PlMH/3UgWmSH/bRb9s11YVoJQ2TMgPlfq/yKKQh0u1nz1v10Xu1Lqle26SembdExJOAMzLzbyPiOGCk3X7SUrZu9UTXEq599JLKKrOgy+uBa4APFUUTwLYKY5JUp1lfvPPDS5qrTJ/6rwPPBb4DkJl3A0+oMqgyfKRNS4Xzw0sqq0xS/0FmPjTzIiKOAdrPLVsxH2nTUrFu9QSXrD+LifExApgYH+OS9WfZny7pKGUGyt0SEb8NjEXEi4E3A9dVG5aket3so5c0vMpcqV8MPADsBt4AXA+8vcqgJElS58qMfj8SER8D/oHabfe9WWZpN0mStKjaJvWIeCnwQeCfqC29enpEvCEzP1N1cJKq5Ux10nAp06d+OfDCzLwHICKeAnwaMKlLA8zV5KThU6ZP/f6ZhF64F7i/ongkLRJXk5OGT9Mr9YhYX3y7JyKuB66m1qf+SuC2RYhNUoWcqU4aPq1uv6+t+/7fgecX3z8APK6yiCR1VbN+8ypWk5PUW02Tema+bjED6VRErAXWrly5stehSH2rVb/5xjWrZr0HzlQnDboyo99PB94KnFa/fWaeV11Y7WXmdcB1k5OTr+9lHFI/a9Vv/vmLX/TINnOv4h0VLw2mMqPftwF/Sm0WuSOVRiOpq9r1mzeaqc5R8dLgKpPU/zMz31t5JJK6bj795q7fLg2uMo+0vSci3hERz46In575qjwySQs2nxXeHBUvDa4yV+pnAb8CvIgf3n7P4rWkPtCsD3zmyrqT/nFHxUuDq0xS/0XgyfXLr0rqH+36wDtd4c1R8dLgKpPUbwfGcRY5qS91uw98Plf3czl6XuqNMkn9icDXIuI24Aczhb1+pE1STRV94AtZv93R81LvlEnq76g8Cknztlh94GWvvh09L/VOmfXUb1mMQCTNz2L0gXdy9e3oeal3yswo911qo90BHgWMAt/PzMdWGZikcrrRB95OuxXd6tseP26Ubz946Kg65t45sN9d6r4yV+qPqX8dEeuAs6sKSFLnFtIHXkazq+yZK/b6K/jRZcHoSHDocD6y3dw7B/a7S9UoM/nMLJm5jYqeUY+IJ0fEn0bENVXUL2l+mvXPj0QcdQV/6Ejy6Ecdw8T4GAFMjI9xyfqzZiVr13KXqlHm9vv6upfLgEl+eDu+rYj4KPAy4P7MfFpd+bnAe4AR4COZeWlm3gtcaFKX+kuzfvu5iXnGwelDfOUdP9+0PvvdpWqUuVJfW/e1Bvgu8PIO2rgCOLe+ICJGgPcDLwHOBC6IiDM7qFPSIlq3eoJL1p911NX3RJMr+HYj75u976x10sKU6VNf0LrqmXlrRJw2p/hs4J7iypyIuIraHwpfXUhbkqrTrN++zMj7uYPiXvjUE/nLnVPOWid1WdOkHhG/22K/zMx3LaDdCeC+utf7gGdFxOOB/w2sjohNmXlJk9g2ABsAVqxYsYAwJC1EmZH3jQbF/eXOKc5/xgQ3fe0BR79LXdTqSv37DcoeDVwIPB5YSFKPBmWZmd8E3thu58zcAmwBmJycLN2/L6n72o28bzYo7qavPcDnL3ZdKKmbmib1zLx85vuIeAxwEfA64Crg8mb7lbQPOLXu9SnA/k4qiIi1wNqVK1cuMBRJVXJQnLR4Wg6Ui4jjI+LdwB3U/gD46cz8H5m50MVdbgPOiIjTI+JRwKuBazupIDOvy8wNy5cvX2AokqrkoDhp8TRN6hGxmVry/S5wVma+MzO/3WkDEXEl8EVgVUTsi4gLM/Nh4C3AduAu4OrM3DOvTyCpr21cs4qx0ZFZZQ6Kk6oRmY27pCPiCLVV2R5m9nPpQa3/u6fTxNbdfn/93Xff3ctQJLXhlLBS90TEzsycbPhes6Q+KCYnJ3PHjh29DkOSpEXRKql3PE2sJEnqTwOb1CNibURsOXjwYK9DkSSpLwxsUnf0uyRJsw1sUpckSbOZ1CVJGhIDm9TtU5ckabaBTer2qUuSNFvbpVclqRecsEbqnEldUt9ptFzrpq27AUzsUgsDe/td0vBqtlzr5u17exSRNBgGNqk7UE4aXi7XKs3PwCZ1B8pJw8vlWqX5sU9dUt/ZuGbVrD51mL1ca9lBdPXbLR8bJQIOPHjIgXcaWiZ1SX1nJtk2StxlB9HN3e7A9KFH3nPgnYaVSV1SX1q3eqJhwm01iK5++0bbtdtHGnQD26fuQDlpaSo7iK7MoDoH3mnYDGxSd6CctDSVHURXZlCdA+80bAY2qUtamjauWcXY6MissvpBdK22a7ePNOjsU5c0UFoNomu1naPftRREZvY6hgWZnJzMHTt29DoMSQs037nenSNeS01E7MzMyUbveaUuqefmO9e7c8RLs9mnLqnn5jvXu3PES7MNbFL3kTZpeMx3rnfniJdmG9ik7iNt0vCY71zvzhEvzTawSV3S8Cj7mFq39pOGlQPlJPVc2cfUurWfNKx8pE2SpAHS6pE2b79LkjQkTOqSJA0J+9QlLVn1s9E1mkYWju6vb1TWzT78uTPkvfCpJ3LT1x4o1d5iza7X6Lh9+8FDjERwOJOJRRrb4GyCR7NPXdKSNHc2urlGlwUEHDqcLcvGRke4ZP1ZXUkm7WJq1V6jfbsZWycxVtV2uziqbrNf2KcuSXM0mo2u3qEjOSt5Nyvr5gx27WJq1d5iza5XJsaq2m4Xh7MJDnBSd0Y5SQvRzVnnulVX2XoabbdYs+t1Ul+VM/s5m2BjA5vUnVFO0kJ0c9a5btVVtp5G2y3W7Hqd1FflzH7OJtjYwCZ1SVqIRrPR1RtdFoyORNuybs5g1y6mVu0t1ux6ZWKsqu12cTiboKPfJS1Rc2ej64fR741myCs7+n2xZtdrdtwWe/S7swk25uh3SZIGiKPfJUlaAkzqkiQNCZO6JElDwqQuSdKQMKlLkjQkTOqSJA0Jk7okSUPCpC5J0pAwqUuSNCT6aprYiHg08CfAQ8DNmfnxHockqQ9s2zXV1elA51Pf3H0aTd8KtWlLpw5MPzJlarN/x+dMS1tfX9mpV7ftmuKd1+7hwPShR8qOG13GsaMjs6a7Xbd6YtGOYaOYlgUcSR75DDPHqdOpcBcSV/179T+fiQbHqP798bFRHnr4MA8eOgLA444b5R1rf+KoOlu1t5jT2FY+TWxEfBR4GXB/Zj6trvxc4D3ACPCRzLw0In4FOJCZ10XEJzLzVe3qd5pYabht2zXFpq27Z62dPTY6wiXrz5rXL8j51Ndon7lGlwUER6233k31cW7bNcXGT97OoSOt2xsbHeH8Z0zwlzunKj+G5z9jgk986b6WMZU5TvONrdXPFmj6M2x2jJp+hpFg8yt+qmGdrdpbyDGv12qa2MVI6j8LfA/485mkHhEjwD8CLwb2AbcBFwAvBz6TmV+JiL/IzF9uV79JXRpuz730c0w1WCN7YnyMz1/8okWpr9k+vTATZycxzVx1NqurU83abtbOfMwntlY/W6Dl8eo09lZ1tntvPse8XqukXvnt98y8NSJOm1N8NnBPZt4LEBFXUUvo+4BTgK/Qor8/IjYAGwBWrFjR/aAl9Y39TX4RNyuvor75tlWFmVg6ialZsur2MexWQm/Vxnz2KVNXp7HP93yp+lzq1UC5CeC+utf7irKtwPkR8QHgumY7Z+aWzJzMzMkTTzyx2kgl9dTJxVVP2fIq6ptvW1WYiaWTmEYiGpZ3+xg2a6ebbcxnn5PHx9rW12nsreps916VepXUGx29zMzvZ+brMvNNDpKTBLBxzSrGRkdmlY2Njjwy4Gox6mu0z1yjy4LRke4ltUbq49y4ZlWtf7rEPhc869RFOYYXPOvUtjGVOU7zja3Vz7bVz7DZMWpmdCSa1tmqvYUc87J6Nfp9H3Bq3etTgP2dVBARa4G1K1eu7GZckvrMzKCibo0ink99jfbp9ej3mX/Ljn6ffNLxi3IMJ590fM9Gv5f52bYa/T5zjDoZ/V6mvaEa/Q5Q9Kl/qm6g3DHUBsr9HDBFbaDcL2fmnk7rdqCcJGkpaTVQrvLb7xFxJfBFYFVE7IuICzPzYeAtwHbgLuDq+SR0SZL0Q4sx+v2CJuXXA9fPt15vv0uSNNvAThObmddl5obly5f3OhRJkvrCwCZ1SZI028Am9YhYGxFbDh482OtQJEnqC4sy+r1KEfEA8PXi5XKgXZZvt80JwH90IbR+Uua4DFrb3ah3vnV0sl+3t/X8HY62PX8bG7bzt6rz50mZ2Xjmtcwcmi9gy0K3AXb0+nP04rgMWtvdqHe+dXSyX7e39fwdjrY9f5u+P1Tnby/O3YG9/d5E06llO9xm2PTyM1fVdjfqnW8dnezX7W09f4ejbc/fpWHRP+/A337vtojYkU0e6pf6neevBpnn78IN25V6N2zpdQDSAnj+apB5/i6QV+qSJA0Jr9QlSRoSJnVJkoaESV2SpCFhUm8hIh4dER+LiA9HxGt6HY/UiYh4ckT8aURc0+tYpE5FxLrid+9fR8TP9zqeQbHkknpEfDQi7o+IO+eUnxsReyPinoi4uCheD1yTma8Hzlv0YKU5Ojl/M/PezLywN5FKR+vw/N1W/O59LfCqHoQ7kJZcUgeuAM6tL4iIEeD9wEuAM4ELIuJM4BTgvmKzw4sYo9TMFZQ/f6V+cwWdn79vL95XCUsuqWfmrcC35hSfDdxTXNk8BFwFvBzYRy2xwxI8Vuo/HZ6/Ul/p5PyNmsuAz2Tmlxc71kFloqqZ4IdX5FBL5hPAVuD8iPgAS296Qw2OhudvRDw+Ij4IrI6ITb0JTWqr2e/ftwLnAK+IiDf2IrBBdEyvA+gT0aAsM/P7wOsWOxipQ83O328C/jJUv2t2/r4XeO9iBzPovFKv2QecWvf6FGB/j2KROuX5q0Hm+dtFJvWa24AzIuL0iHgU8Grg2h7HJJXl+atB5vnbRUsuqUfElcAXgVURsS8iLszMh4G3ANuBu4CrM3NPL+OUGvH81SDz/K2eC7pIkjQkltyVuiRJw8qkLknSkDCpS5I0JEzqkiQNCZO6JElDwqQuSdKQMKlr4ERERsTlda/fFhHv7FLdV0TEK7pRV5t2XhkRd0XETRW2cWVE3BER/72qNhq0+dtzXn+hwrbOq1smucz24xHx5rrXL4iIT1UQ1/fmud86V9fTQpnUNYh+AKyPiBN6HUi9YgnJsi4E3pyZL6wolh8DnpOZP5mZf1Ryn26sBTErqWfmc7pQZ0OZeW1mXtrBLuPAm9tt1EPrqC09WlqXfmYaIiZ1DaKHgS3AUVegc6+0Z66aiquyWyLi6oj4x4i4NCJeExFfiojdEfGUumrOiYi/K7Z7WbH/SERsjojbiqvfN9TVe1NE/AWwu0E8FxT131ksI0lE/C7wPOCDEbG5wT6/Vexze0RcWpQ9PSL+vmj7ryLicUX5zRFxWfE5/jEifqao5rPAEyLiKxHxM232//2IuAW4qHj9RxFxa3En4ZkRsTUi7o6Id9fFuC0idkbEnojYUJRdCowVbX58zvGP4vjdWXy2V9Udv5sj4pqI+FpEfDwiYqa+iPhqEfMfNDhOr42I99X93N8bEV+IiHub3G25FHhKEd/Mcf/RJm0/ozhfdkbE9og4qUH7TyyO5e3F13PmvD/rTkBEvC8iXtvosxX7ngdsLuJ7SvF1QxHD30XEU+s+6x9G7S7PZQ0+p5ayzPTLr4H6Ar4HPBb4F2A58DbgncV7VwCvqN+2+PcFwAHgJOBYYAr4veK9i4A/rtv/Bmp/8J5BbbGJHwE2AG8vtjkW2AGcXtT7feD0BnGeDPwrcCK1FRE/B6wr3rsZmGywz0uALwDHFa+PL/69A3h+8f3/qov3ZuDy4vtfAP62+P404M66elvt/yd1290MXFZ3XPbXHbN9wOPnxDUG3FlX/r25P6vi3/OBvwFGgCcWx+Wk4vgdpLaIxzJqU4g+Dzge2MsPZ70cb3CsXgu8r+7n9smijjOprc89d/u5x6RZ26PFz+DEYrtXAR9tUN8ngN8svh8Bljc45z5Vt/37ipgbfjaOPndvBM4ovn8W8Lm67T4FjPT6/6Jf/fflrRsNpMz8TkT8OfAbwHTJ3W7LzG8ARMQ/UbuahdoVdv1t8Ksz8whwd0TcCzwV+HngJ+uuAJdTS/oPAV/KzH9u0N4zgZsz84GizY8DPwtsaxHjOcCfZeaDxef8VkQsp/aL/5Zim49RS2Azthb/7qSWuGYpsf8n5uwys5jGbmBP3TG7l9pqWt8EfiMifrHY7lRqx+KbLT7X84ArM/Mw8O/FnYFnAt+hdvz2FW18pfgMfw/8J/CRiPg0tSTWzrbi5/bViHhiie1p0vYB4GnA3xQX7iPANxrs+yLgvwEUn+tgyTa/Q5vPFhE/CjwH+GQRA9T+sJrxyaJNaRaTugbZHwNfBv6sruxhim6l4lbqo+re+0Hd90fqXh9h9v+FuQsiJLU1n9+amdvr34iIF1C7Um+k0TrR7USD9tuZ+RyHmd//6bnx1x+XucfsmOIznwM8OzMfjIibqd3NaKXVsahv4zBwTGY+HBFnAz9HbdWut1BLoq3U11P22B/VdrHvnsx8dsk6mnnkXCz8CEDJz7YMOJCZT29Sd7NzTkucfeoaWJn5LeBqaoPOZvwL8Izi+5dTu5XaqVdGxLKo9bM/mdqt0u3AmyJiFCAifjwiHt2mnn8Anh8RJ0RtEN0FwC1t9vks8GsRcVzRzvGZeRD4dl1/+a+UqOcRC92/geXAt4uE/lTgv9a9d2jmGM1xK/CqqI1NOJHaHYsvNWuguFJdnpnXA78JPH0B8c74LvCYEtvtBU6MiGcXsYxGxE802O5G4E3FNiMR8dg5738dODMiji3ulvxcsW2zz/ZIfJn5HeCfI+KVxT4RET9V9oNq6TKpa9BdDtSPgv8wtUT6JWr9kPO5otlLLel9BnhjZv4n8BHgq8CXI+JO4EO0uSoubltvAm4Cbge+nJl/3WafG6jd/t5R3A5+W/HWr1IbRHUHtSTwvzr8TAvdv94N1K7Y7wDeRe1W+YwtwB0zA+Xq/BW1fv3bqY0t+K3M/LcWbTwG+FTRxi00GBTZqcz8JvD5YrDeUQMU67Z7CHgFcFlE3A58hdqt8LkuAl4YEbupdX3MSvyZeR+1PzrvAD4O7CreavbZrgI2RsSu4g/K1wAXFjHsofZHqtSSS69KkjQkvFKXJGlImNQlSRoSJnVJkoaESV2SpCFhUpckaUiY1CVJGhImdUmShoRJXZKkIfH/Ac4HNasG/TPrAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 576x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#@title\n",
    "clustsizes = []\n",
    "for motif in motifs:\n",
    "    for clust in clusters[motif].values():\n",
    "        clustsizes.append(len(clust))\n",
    "x, y = np.unique(clustsizes, return_counts=True)\n",
    "        \n",
    "fig = plt.figure(figsize=(8,4))\n",
    "gs = gridspec.GridSpec(1, 1)\n",
    "ax1 = plt.subplot(gs[0, 0])\n",
    "ax1.scatter(x, y)\n",
    "ax1.set_xscale(\"log\")\n",
    "ax1.set_yscale(\"log\")\n",
    "\n",
    "fig.text(0.055, 0.5, \"Number of clusters\", va='center', rotation='vertical')\n",
    "plt.xlabel(\"Number of conformations in the cluster\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "name": "filtering-clustering.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
